### PR TITLE
feat: Strict preflight with handler

### PR DESCRIPTION
**Pull request by codyzu**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

**This is an alternative implementation to #95 after some discussion with @mcollina.**

This implementation refactors the preflight response logic into cors route handler. This has the advantage that users can override the CORS options route handler (it is a  low priority `*` route) for specific routes. This fits better with the fastify routing table model.

This obsoletes the [`preflightContinue` option](https://github.com/fastify/fastify-cors/tree/bfb2cf63ace5253ec599d46f7e7eac8ba0edc4ec#options). Previously, this option stopped the logic in the hook that replies and "short circuited" the routing of the request.

This solution will allow users to simply add options routes, if they so choose, and have control over the responses. The `strictPreflight` option now has 2 functionalities when set to true:
1. stops adding preflight headers if the origin and access-control-request-method headers are missing
1. replies with 400 if those headers are missing (this can now be overridden by adding an options route).

#### From the previous PR:

This is a new feature that attempts to resolve the issue revealed in #88 .

This PR introduces a new option `strictPreflight` that will enforce the origin and access-control-request-method headers on preflight options requests.

Instead, it will result in a 400 response if preflight requests do not have the required headers. This aligns the fastify-cors plugin with the now superseded [w3c specification describing the preflight workflow](https://www.w3.org/TR/2020/SPSD-cors-20200602/#resource-preflight-requests).

The superseded spec states that when the headers are missing the preflight processing should:

>  ... terminate this set of steps. The request is outside the scope of this specification.

The new [fetch spec](https://fetch.spec.whatwg.org/#cors-preflight-request) does not indicate what the behavior should be if the headers are missing.

Finally, while investigating this issue, I compared the current behavior of fastify-cors with express cors. **Fastify-cors behaves in the same way as express cors.** This is the motivation for introducing the new option that would allow consumers of the plugin to opt-in to stricter handling of the preflight headers.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -25,15 +25,34 @@ function fastifyCors (fastify, opts, next) {
...
 
   const isOriginFalsy = !origin
   const isOriginString = typeof origin === 'string'
   const isOriginFunction = typeof origin === 'function'
 
   if (preflight === true) {
-    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => reply.send())
+    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
+      if (origin === false) {
+        reply.send()
```
Comment (codyzu) [inline index.js:39]: This is how the plugin previously worked.

IMO it would be better:
```javascript
reply.code(404).type('text/html').send('Not Found')
```

But I wanted to keep compatibility :thinking:

Comment (mcollina) [reply index.js:39]: Go for it.

File: index.js [hunk 2]
```diff
@@ -58,8 +77,8 @@ function fastifyCors (fastify, opts, next) {
         )
       }
 
-      if (req.raw.method === 'OPTIONS' && preflight === true) {
-        // preflight
+      // Handle preflight headers (if strict mode is enabled, then the valid preflight headers must exit)
+      if (req.raw.method === 'OPTIONS' && preflight === true && (strictPreflight === false || (req.headers.origin && req.headers['access-control-request-method']))) {
```
Comment (mcollina) [inline index.js:81]: I think we should move this if inside the route.

File: test/index.test.js [hunk 3]
```diff
@@ -10,7 +10,7 @@ test('Should shortcircuits preflight requests', t => {
   const fastify = Fastify()
   fastify.register(cors)
 
-  fastify.options('/', (req, reply) => {
+  fastify.get('/', (req, reply) => {
```
Comment (mcollina) [inline test/index.test.js:13]: why this change?

File: test/index.test.js [hunk 4]
```diff
@@ -37,7 +37,7 @@ test('Should add access-control-allow-headers to response if preflight req has a
...
```
Comment (mcollina) [inline test/index.test.js:40]: why this change?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/index.test.js (hunk 3 above)
Comment (codyzu) [reply test/index.test.js:13]: Leaving this as `options`, would override the route added inside fastify-cors and that is not the intention of the test.

I changed these to `get` as just kind of a hacky fix to the tests, but perhaps it would be better just to remove these routes... That more sense, if the route is not "short-circuited" it would finish with a 404. I'll do that.

File: index.js (hunk 2 above)
Comment (codyzu) [reply index.js:81]: I started that way, but then decided to leave the headers stuff in the hook to keep the behavior similar to the previous `preflightContinue` behavior. The previous behavior would still add the headers, but leave the application to reply.
:point_up_2: 
I guess I took the perspective of minimal functional difference, but I see your point that it is just not the "fastify" way to do plugins and probably makes it more complicated to understand.

I'll see what it looks like if I move this header handling into the route.

File: test/index.test.js [hunk 5]
```diff
@@ -451,10 +439,11 @@ test('Should not add cors headers other than `vary` when origin is false', t =>
...
     t.error(err)
     delete res.headers.date
-    t.strictEqual(res.statusCode, 200)
-    t.strictEqual(res.payload, '')
+    t.strictEqual(res.statusCode, 404)
+    t.strictEqual(res.payload, 'Not Found')
     t.deepEqual(res.headers, {
-      'content-length': '0',
+      'content-length': '9',
+      'content-type': 'text/plain',
```
Comment (mcollina) [inline test/index.test.js:446]: I don't understand this change. Can you please it a bit? It seems `origin: false` with `preflight: true` does not make sense anymore. I would not add the route.

File: test/index.test.js [hunk 6]
```diff
@@ -694,10 +683,6 @@ test('Allow only request from with specific methods', t => {
   const fastify = Fastify()
   fastify.register(cors, { methods: ['GET', 'POST'] })
 
-  fastify.options('/', (req, reply) => {
-    reply.send('ok')
-  })
```
Comment (mcollina) [inline test/index.test.js:699]: why was this removed?

File: index.js [hunk 7]
```diff
@@ -25,15 +25,57 @@ function fastifyCors (fastify, opts, next) {
...
   const isOriginString = typeof origin === 'string'
   const isOriginFunction = typeof origin === 'function'
 
   if (preflight === true) {
-    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => reply.send())
+    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
+      if (origin === false) {
+        reply.code(404).type('text/plain').send('Not Found')
+        return
+      }
```
Comment (mcollina) [inline index.js:41]: I think we should not add the route if `!origin && preflight`.

File: index.js [hunk 8]
```diff
@@ -25,15 +25,57 @@ function fastifyCors (fastify, opts, next) {
     allowedHeaders: null,
     maxAge: null,
     preflight: true,
-    hideOptionsRoute: true
+    hideOptionsRoute: true,
+    strictPreflight: false
```
Comment (mcollina) [inline index.js:29]: I would flip this to `true` and bump the major. This is a substantial change already.

wdyt @jsumners?

Comment (jsumners) [reply index.js:29]: Sounds good to me.

File: test/index.test.js (hunk 6 above)
Comment (codyzu) [reply test/index.test.js:699]: The route never gets called by the test. I removed it to keep the tests concise. I only kept routes that actually get called. Do you prefer I put it back?

Comment (mcollina) [reply test/index.test.js:699]: I have a feeling this was called by the following `.inject()` call. why wasn't it called?

File: test/index.test.js [hunk 9]
```diff
@@ -751,3 +736,86 @@ test('Allow only request from with specific headers', t => {
...
+
+  const fastify = Fastify()
+  fastify.register(cors, {
+    strictPreflight: true
+  })
+
+  fastify.get('/', (req, reply) => {
+    t.fail('we should not be here')
+    reply.send('ok')
+  })
```
Comment (mcollina) [inline test/index.test.js:751]: I think this route is not needed.

File: test/index.test.js [hunk 10]
```diff
@@ -751,3 +736,86 @@ test('Allow only request from with specific headers', t => {
...
```
Comment (mcollina) [inline test/index.test.js:777]: I think this route is not needed

File: test/index.test.js [hunk 11]
```diff
@@ -751,3 +736,86 @@ test('Allow only request from with specific headers', t => {
...
+
+test('Should shortcircuit preflight requests with origin and access control method headers when strictPreflight is enabled', t => {
+  t.plan(4)
+
+  const fastify = Fastify()
+  fastify.register(cors, { strictPreflight: true })
+
+  fastify.get('/', (req, reply) => {
+    t.fail('we should not be here')
+  })
```
Comment (mcollina) [inline test/index.test.js:800]: I think this route is not needed

**Review by mcollina — CHANGES_REQUESTED** [review]
Could you split the tests into 2 files? One for the preflight/OPTIONS request?

File: test/index.test.js (hunk 6 above)
Comment (codyzu) [reply test/index.test.js:699]: You are correct, I confused this with the other tests.

Here, in the previous implementation, the request was handled and terminated in the onRequest hook since `preflightContinue` was the default value of false.

In my new implementation, when overriding an OPTIONS route, the preflight headers will not be added by the plugin (the `*` route handler will never execute). In order to test the `Access-Control-Allow-Methods` header, the route must _not_ be handled by a "user" route. Instead, it should be left to be handled by the route defined in this plugin.

Does that make more sense?

Comment (mcollina) [reply test/index.test.js:699]: Yes it does

File: index.js (hunk 7 above)
Comment (codyzu) [reply index.js:41]: This is more complicated than I first realized.

In the original implementation, `origin` could be the result of a callback. As I see it, this means a few things:
1) My implementation here is not functionally equivalent to the previous implementation (I'm not calling the callback)
2) Makes not adding the route not feasible because if origin is a callback, it should be called on every request.

Comment (mcollina) [reply index.js:41]: Then call `origin`.

Comment (codyzu) [reply index.js:41]: I could call the origin callback in the hook and decorate the request with the result :thinking: Then check the value here and return 404 if it is false.

These options just keep getting more complicated :facepalm:

Comment (codyzu) [reply index.js:41]: Is it weird if it gets called twice during the same request? Once in the hook when adding the CORS headers and once in the route when adding the preflight stuff

Comment (codyzu) [reply index.js:41]: I could decorate the request and assign the result of the origin function to the request in the hook. That way it will only be called once per request. This is complicated but makes sense if the origin function is calling some external service or other weird behavior since it can return a promise.

File: test/index.test.js (hunk 5 above)
Comment (codyzu) [reply test/index.test.js:446]: See our discussion above. Origin can be a function and dynamically determined for each request.

So you are correct that it does not make sense and I'm returning 404. This is clearly a bit hacky... :thinking: thinking about better solutions.

File: index.js (hunk 7 above)
Comment (codyzu) [reply index.js:42]: @mcollina I went with decorating the request with a boolean determining if the origin is allowed. This avoids calling a possibly asynchronous function (or promise) twice on every request.

Comment (mcollina) [reply index.js:42]: good job! I like this.

File: index.js (hunk 8 above)
Comment (mcollina) [reply index.js:29]: @codyzu ^^

File: test/index.test.js (hunk 5 above)
Comment (codyzu) [reply test/index.test.js:296]: @mcollina is this change ok now with the new implementation that does allow for `origin: false` with `preflight: true` (will return 404).

**Comment by codyzu** [comment]
@mcollina @jsumners this should have all of the open points resolved. I tested it manually in Chrome with the new default `strictPreflight` option (now true) and it worked as expected. Let me know if you see anything else... I bumped the major version in my branch too.

**Review by mcollina — APPROVED** [review]
lgtm
